sdks/python/apache_beam/testing/benchmarks/nexmark/queries/query6.py

```diff
@@ -0,0 +1,93 @@
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""
+Query 6, 'Average Selling Price by Seller'. Select the average selling price
+over the last 10 closed auctions by the same seller. In CQL syntax::
+
+  SELECT Istream(AVG(Q.final), Q.seller)
+  FROM (SELECT Rstream(MAX(B.price) AS final, A.seller)
+    FROM Auction A [ROWS UNBOUNDED], Bid B [ROWS UNBOUNDED]
+    WHERE A.id=B.auction
+      AND B.datetime < A.expires AND A.expires < CURRENT_TIME
+    GROUP BY A.id, A.seller) [PARTITION BY A.seller ROWS 10] Q
+  GROUP BY Q.seller;
+"""
+
+from __future__ import absolute_import
+from __future__ import division
+
+import apache_beam as beam
+from apache_beam.testing.benchmarks.nexmark.models import seller_price
+from apache_beam.testing.benchmarks.nexmark.queries import nexmark_query_util
+from apache_beam.testing.benchmarks.nexmark.queries import winning_bids
+from apache_beam.transforms import trigger
+from apache_beam.transforms import window
+
+
+def load(events, metadata=None):
+  # find winning bids for each closed auction
+  return (
+      events
+      # find winning bids
+      | beam.Filter(nexmark_query_util.auction_or_bid)
+      | winning_bids.WinningBids()
+      # (auction_bids -> (aution.seller, bid)
+      | beam.Map(lambda auc_bid: (auc_bid.auction.seller, auc_bid.bid))
+      # calculate and output mean as data arrives
+      | beam.WindowInto(
+          window.GlobalWindows(),
+          trigger=trigger.Repeatedly(trigger.AfterCount(1)),
+          accumulation_mode=trigger.AccumulationMode.ACCUMULATING,
+          allowed_lateness=0)
+      # | beam.combiners.Count.Globally())
+      | beam.CombinePerKey(MovingMeanSellingPriceFn(10))
+      | beam.Map(lambda t: seller_price.SellerPrice(t[0], t[1])))
+
+
+class MovingMeanSellingPriceFn(beam.CombineFn):
+  """
+  Combiner to keep track of up to max_num_bids of the most recent wining
+  bids and calculate their average selling price.
+  """
+  def __init__(self, max_num_bids):
+    self.max_num_bids = max_num_bids
+
+  def create_accumulator(self):
+    return []
+
+  def add_input(self, accumulator, element):
+    accumulator.append(element)
+    new_accu = sorted(accumulator, key=lambda bid: (bid.date_time, bid.price))
+    if len(new_accu) > self.max_num_bids:
```

Comment:
should this be a while loop?

Reply:
the length of the `new_accu` should never be larger than `max_num_bids` by more than 1 because each time 1 input is added, this if statement is checked and oldest element will be dropped starting from the first time it went over, so if would be ok to have if instead of while.
But logically it would be while if input is not always size 1.